drivers/adc/adc_stm32f4.c

```diff
@@ -0,0 +1,501 @@
+			GPIOport = GPIOC;
+			gpioCfg.Pin = GPIO_PIN_5;
+			cfg.Channel = ADC_CHANNEL_15;
+			break;
+
+		case adcChannel_vref:
+			cfg.Channel = ADC_CHANNEL_VREFINT;
+			ADC->CCR |= ADC_CCR_TSVREFE;
+			break;
+
+		case adcChannel_temp:
+			cfg.Channel = ADC_CHANNEL_TEMPSENSOR;
+			break;
+
+		case adcChannel_vbat:
+			cfg.Channel = ADC_CHANNEL_VBAT;
+			ADC->CCR |= ADC_CCR_VBATE;
+			break;
+
+		default:
+			// none
+			break;
+	}
+	if(GPIOport) HAL_GPIO_Init(GPIOport, &gpioCfg);
+
+	if (HAL_ADC_ConfigChannel(phadc, &cfg) != HAL_OK) {
+		SYS_LOG_ERR("config adc channel failed");
+		return stm32adcError_adcHALconfigChannel;
+	}
+	SYS_LOG_DBG("internal adc channel %u has been configured",cfg.Channel);
+
+	return stm32adcError_None;
+}
+
+int adc_stm32f4_init(struct device *dev)
+{
+	const struct adc_config *config = dev->config->config_info;
+	struct adc_drvData *drvData = dev->driver_data;
+	adc_channel_index_t i;
+	uint32_t r;
+	uint32_t activeChannels;
+
+	SYS_LOG_INF("init adc%u", (unsigned int)config->adcDevNum);
+
+	switch(config->adcDevNum) {
+		#ifdef CONFIG_ADC_0
+		case 0:
+			__HAL_RCC_ADC1_CLK_ENABLE();
+			drvData->hadc.Instance = ADC1;
+			break;
+		#endif
+
+		#ifdef CONFIG_ADC_1
+		case 1:
+			__HAL_RCC_ADC2_CLK_ENABLE();
+			drvData->hadc.Instance = ADC2;
+			break;
+		#endif
+		#ifdef CONFIG_ADC_2
+		case 2:
+			__HAL_RCC_ADC3_CLK_ENABLE();
+			drvData->hadc.Instance = ADC3;
+			break;
+		#endif
+		default:
+			SYS_LOG_ERR("unknown ADC unit");
+			return stm32adcError_UnknownADCunit;
+	}
+
+	drvData->hadc.Init.ClockPrescaler = ADC_CLOCK_SYNC_PCLK_DIV2;
+	drvData->hadc.Init.Resolution = ADC_RESOLUTION_12B;
+	drvData->hadc.Init.ScanConvMode = ENABLE;
+	drvData->hadc.Init.ContinuousConvMode = ENABLE;
+	drvData->hadc.Init.DiscontinuousConvMode = DISABLE;
+	drvData->hadc.Init.ExternalTrigConvEdge = ADC_EXTERNALTRIGCONVEDGE_NONE;
+	drvData->hadc.Init.ExternalTrigConv = ADC_SOFTWARE_START;
+	drvData->hadc.Init.DataAlign = ADC_DATAALIGN_RIGHT;
+	drvData->hadc.Init.DMAContinuousRequests = DISABLE;
+	drvData->hadc.Init.EOCSelection = ADC_EOC_SINGLE_CONV;
+	drvData->hadc.Init.NbrOfConversion = 0;
+
+	activeChannels = config->activeChannels;
+	while(activeChannels) {
+		if(activeChannels & 1) drvData->hadc.Init.NbrOfConversion++;
+		activeChannels >>= 1;
+	}
+	SYS_LOG_INF("use %u multiplexed channels", (unsigned int) drvData->hadc.Init.NbrOfConversion);
+
+	// start the adc
+	if (HAL_ADC_Init(&drvData->hadc) != HAL_OK) {
+		SYS_LOG_ERR("HAL ADC init failed");
+		return stm32adcError_adcHALinit;
+	}
+
+	// setup the sequencer for the channels
+	i = 0;
+	r = 1;
+	activeChannels = config->activeChannels;
+	while(activeChannels) {
+		if(activeChannels & 1) {
+			if(adc_stm32f4_cfgChannel(&drvData->hadc, (uint32_t) i, r)) {
+				SYS_LOG_ERR("activate adc channel %u failed", i);
+				return stm32adcError_configChannel;
+			} else {
+				SYS_LOG_INF("activate adc channel %u", i);
+			}
+			r++;
+		}
+		activeChannels >>= 1;
+		i++;
+	}
```

Comment:
To implement this we need to check/setup the adc sequencer at each adc run. This will cost some computing time. Think usually the user has a static setup which does not change at runtime.

A limiting factor to the sequencing structure is, that you can not define a real delay between the conversation of the channels. This Zephyr feature is not supported by the stm32f4 hardware. Of course we can try to add this via software. But is this really needed?

Reply:
Unfortunately this is not my setup, I need to reconfigure the ADC sequence at runtime to address some complex needs.
I do not use the conversion delay from the sequencing structure so this is not really a problem for me.